Unable to initialize dev mode handler. A handler is already present

This seems to happen in Vaadin 22 when deploying a Spring Boot app as a war.
```
Caused by: java.lang.IllegalStateException: Unable to initialize dev mode handler. A handler is already present: com.vaadin.base.devserver.WebpackHandler@7e6364ec
	at com.vaadin.base.devserver.DevModeHandlerManagerImpl.setDevModeHandler(DevModeHandlerManagerImpl.java:64) ~[vaadin-dev-server-9.0.0.beta4.jar:na]
	at com.vaadin.base.devserver.DevModeHandlerManagerImpl.initDevModeHandler(DevModeHandlerManagerImpl.java:79) ~[vaadin-dev-server-9.0.0.beta4.jar:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$DevModeServletContextListener.failFastContextInitialized(VaadinServletContextInitializer.java:457) ~[vaadin-spring-19.0.0.beta3.jar:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$FailFastServletContextListener.contextInitialized(VaadinServletContextInitializer.java:189) ~[vaadin-spring-19.0.0.beta3.jar:na]
	... 33 common frames omitted
```

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
@@ -429,6 +429,9 @@ public class VaadinServletContextInitializer
                                 + "run the build-frontend maven goal) or "
                                 + "include the vaadin-dev-server dependency");
             }
+            if (isDevModeAlreadyStarted(event.getServletContext())) {
+                return;
+            }
 
             Set<String> basePackages;
             if (isScanOnlySet()) {
@@ -495,6 +498,21 @@ public class VaadinServletContextInitializer
             return customScanOnly != null && !customScanOnly.isEmpty();
         }
 
+        private boolean isDevModeAlreadyStarted(ServletContext servletContext) {
+            if (devModeHandlerManager != null) {
+                VaadinServletContext vaadinContext = new VaadinServletContext(
+                        servletContext);
+                if (devModeHandlerManager
+                        .isDevModeAlreadyStarted(vaadinContext)) {
+                    if (getLogger().isDebugEnabled()) {
+                        getLogger().debug(
+                                "Skipped DevModeHandler initialization as it has been already initialized");
+                    }
+                    return true;
+                }
+            }
+            return false;
+        }
     }
 
     private class WebComponentServletContextListener

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
@@ -63,13 +63,10 @@ import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.di.LookupInitializer;
 import com.vaadin.flow.internal.DevModeHandler;
 import com.vaadin.flow.internal.DevModeHandlerManager;
-import com.vaadin.flow.internal.ReflectTools;
 import com.vaadin.flow.router.HasErrorParameter;
-import com.vaadin.flow.router.NotFoundException;
 import com.vaadin.flow.router.Route;
 import com.vaadin.flow.router.RouteAlias;
 import com.vaadin.flow.router.RouteConfiguration;
-import com.vaadin.flow.router.RouteNotFoundError;
 import com.vaadin.flow.server.AmbiguousRouteConfigurationException;
 import com.vaadin.flow.server.InvalidRouteConfigurationException;
 import com.vaadin.flow.server.RouteRegistry;
@@ -429,7 +426,11 @@ public class VaadinServletContextInitializer
                                 + "run the build-frontend maven goal) or "
                                 + "include the vaadin-dev-server dependency");
             }
-            if (isDevModeAlreadyStarted(event.getServletContext())) {
+            if (devModeHandlerManager.getDevModeHandler() != null) {
+                /*
+                 * If a Spring Boot app is deployed as a war, the initializers
+                 * have already been run and should not be run again here
+                 */
                 return;
             }
 
@@ -498,21 +499,6 @@ public class VaadinServletContextInitializer
             return customScanOnly != null && !customScanOnly.isEmpty();
         }
 
-        private boolean isDevModeAlreadyStarted(ServletContext servletContext) {
-            if (devModeHandlerManager != null) {
-                VaadinServletContext vaadinContext = new VaadinServletContext(
-                        servletContext);
-                if (devModeHandlerManager
-                        .isDevModeAlreadyStarted(vaadinContext)) {
-                    if (getLogger().isDebugEnabled()) {
-                        getLogger().debug(
-                                "Skipped DevModeHandler initialization as it has been already initialized");
-                    }
-                    return true;
-                }
-            }
-            return false;
-        }
     }
 
     private class WebComponentServletContextListener

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
@@ -87,6 +87,9 @@ public class VaadinServletContextInitializerTest {
             throws Exception {
         initDefaultMocks();
 
+        Mockito.when(
+                devModeHandlerManager.isDevModeAlreadyStarted(Mockito.any()))
+                .thenReturn(false);
         Mockito.when(devModeHandlerManager.getHandlesTypes())
                 .thenReturn(new Class<?>[0]);
 
@@ -102,6 +105,32 @@ public class VaadinServletContextInitializerTest {
         // should have been called exactly one time.
         Mockito.verify(devModeHandlerManager).initDevModeHandler(Mockito.any(),
                 Mockito.any(VaadinContext.class));
+    }
+
+    @Test
+    public void onStartup_devModeAlreadyInitialized_devModeInitializationSkipped()
+            throws Exception {
+        initDefaultMocks();
+
+        Mockito.when(
+                devModeHandlerManager.isDevModeAlreadyStarted(Mockito.any()))
+                .thenReturn(true);
+        Mockito.when(devModeHandlerManager.getHandlesTypes())
+                .thenReturn(new Class<?>[0]);
+
+        Mockito.when(appConfig.enableDevServer()).thenReturn(true);
+
+        VaadinServletContextInitializer vaadinServletContextInitializer = getStubbedVaadinServletContextInitializer();
+
+        // Simulate Spring context start only
+        vaadinServletContextInitializer.onStartup(servletContext);
+
+        // In our case, we want to check if Dev Mode has been started within
+        // onStartup() call, that means DevModeInitializer.initDevModeHandler()
+        // should not have been called.
+        Mockito.verify(devModeHandlerManager, Mockito.never())
+                .initDevModeHandler(Mockito.any(),
+                        Mockito.any(VaadinContext.class));
     }
 
     @Test

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
@@ -29,6 +29,7 @@ import org.springframework.core.env.Environment;
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.function.DeploymentConfiguration;
+import com.vaadin.flow.internal.DevModeHandler;
 import com.vaadin.flow.internal.DevModeHandlerManager;
 import com.vaadin.flow.router.BeforeEnterEvent;
 import com.vaadin.flow.router.ErrorParameter;
@@ -87,9 +88,8 @@ public class VaadinServletContextInitializerTest {
             throws Exception {
         initDefaultMocks();
 
-        Mockito.when(
-                devModeHandlerManager.isDevModeAlreadyStarted(Mockito.any()))
-                .thenReturn(false);
+        Mockito.when(devModeHandlerManager.getDevModeHandler())
+                .thenReturn(null);
         Mockito.when(devModeHandlerManager.getHandlesTypes())
                 .thenReturn(new Class<?>[0]);
 
@@ -112,9 +112,9 @@ public class VaadinServletContextInitializerTest {
             throws Exception {
         initDefaultMocks();
 
-        Mockito.when(
-                devModeHandlerManager.isDevModeAlreadyStarted(Mockito.any()))
-                .thenReturn(true);
+        DevModeHandler devModeHandler = Mockito.mock(DevModeHandler.class);
+        Mockito.when(devModeHandlerManager.getDevModeHandler())
+                .thenReturn(devModeHandler);
         Mockito.when(devModeHandlerManager.getHandlesTypes())
                 .thenReturn(new Class<?>[0]);
 

